cmd/mt-index-cat/out/template_functions.go

```diff
@@ -0,0 +1,70 @@
+package out
+
+import (
+	"math/rand"
+	"strings"
+)
+
+// random choice between replacing a node with a wildcard, a char with a wildcard, and passthrough
+func pattern(in string) string {
+	mode := rand.Intn(3)
+	if mode == 0 {
+		// in this mode, replaces a node with a wildcard
+		parts := strings.Split(in, ".")
+		parts[rand.Intn(len(parts))] = "*"
+		return strings.Join(parts, ".")
+	} else if mode == 1 {
+		// randomly replace chars with a *
+		// note that in 1/5 cases, nothing happens
+		// and otherwise, sometimes valid patterns are produced,
+		// but it's also possible to produce patterns that won't match anything (if '.' was taken out)
+		chars := rand.Intn(5)
+		pos := rand.Intn(len(in) - chars)
+		return in[0:pos] + "*" + in[pos+chars:]
+	}
+	// mode 3: do nothing :)
+	return in
+}
+
+// age returns age compared to time of query
+func age(in int64) int64 {
+	return QueryTime - in
+}
+
+// roundDuration rounds a second-specified duration for rough classification
+func roundDuration(in int64) int64 {
+	if in <= 10 { // 10s -> don't round
+		return in
+	} else if in <= 60 { // 1min -> round to 10s
+		return round(in, 10)
+	} else if in <= 600 { // 10min -> round to 1min
+		return round(in, 60)
+	} else if in <= 3600 { // 1h -> round to 10min
+		return round(in, 600)
+	} else if in <= 3600*24 { // 24h -> round to 1h
+		return round(in, 3600)
+	} else if in <= 3600*24*7 { // 7d -> round to 1d
+		return round(in, 3600*24)
+	} else if in <= 3600*24*30 { // 30d -> round to 7d
+		return round(in, 3600*24*7)
+	}
+	// default to rounding to months
+	return round(in, 3600*24*30)
+}
+
+// round rounds number d to the nearest r-boundary
+func round(d, r int64) int64 {
+	neg := d < 0
```

Comment:
round appears to be prepared to deal with negative numbers as if they were positive numbers, but the list of `else if`s in `roundDuration` appears to assume all numbers are positive. I think for consistency it would be better to either deal with negative numbers correctly or not.

Reply:
roundDuration rounds durations. it seems to be implied to me that durations can't be negative. 
i see your point though but i would rather not cripple a utility function because it's currently only being called by roundDuration.
we can argue both ways on this but it's not worth the time.

Reply:
of course a duration can be negative.  The sign tells you if the duration is the amount of time before or after the reference time.

Via a comment, at https://github.com/grafana/metrictank/blob/cda12b79a4c020aee9a6065fe10a83694358bded/cmd/mt-index-cat/main.go#L192

You state that age cant be negative, but that is not true.  If the local time where this code is running is wrong, then age can easily be negative.  Likewise, if the localtime on the server that is writing to the index is wrong the age can also end up negative.

Both of these are problems that we should not be masking.

Reply:
But looking at the code, negatives already look to be handled.